pkg/syscallfilter/syscallfilter_linux.go

```diff
@@ -0,0 +1,193 @@
+	"SignalExit":   strace.SignalExit,
+	"Exit":         strace.Exit,
+	"SignalStop":   strace.SignalStop,
+	"NewChild":     strace.NewChild,
+}
+
+var allActions = map[string]interface{}{
+	// Error will end the strace, expeditiously, unless the value is 0
+	"error": nil,
+	// log will log the record
+	"log": nil,
+}
+
+func findEvent(events []*event, n string) *event {
+	for _, e := range events {
+		m, err := regexp.MatchString(e.Pat, n)
+		if err != nil {
+			log.Fatalf("Can't happen: %q is bad: %v", e.Pat, err)
+		}
+		if m {
+			return e
+		}
+	}
+	return nil
+}
+
+// eventName returns an event name. There should never be an event name
+// we do not known and, if we encounter one, we panic.
+func eventName(r *strace.TraceRecord) string {
+	// form up a reasonable name for a system call.
+	// If there is no name, then it will be Exxxx or Xxxxx, where x
+	// is the system call number as %04x.
+	// Note that users can specify this: E0x0000, for example
+	var sysname string
+
+	switch r.Event {
+	case strace.SyscallEnter, strace.SyscallExit:
+		var err error
+		if sysname, err = strace.ByNumber(uintptr(r.Syscall.Sysno)); err != nil {
+			sysname = fmt.Sprintf("%04x", r.Syscall.Sysno)
+		}
+	}
+	switch r.Event {
+	case strace.SyscallEnter:
+		return "E" + sysname
+	case strace.SyscallExit:
+		return "X" + sysname
+	case strace.SignalExit:
+		return fmt.Sprintf("SignalExit")
+	case strace.Exit:
+		return fmt.Sprintf("Exit")
+	case strace.SignalStop:
+		return fmt.Sprintf("SignalStop")
+	case strace.NewChild:
+		return fmt.Sprintf("NewChild")
+	}
+	log.Panicf("Unknown event %#x from record %v", r.Event, r)
+	return ""
+}
+
+func (c *Cmd) handleEvent(t strace.Task, r *strace.TraceRecord, e []*event) error {
+	// All attempts to use defer for printing got ... weird.
+	var ret error
+	n := eventName(r)
+	act := findEvent(e, n)
+	if act == nil {
+		return nil
+	}
+	switch act.Action {
+	case "error":
+		ret = fmt.Errorf("%v", act.Value)
+		if c.Log != nil {
+			fmt.Fprintf(c.Log, "%v act %v error %v\n", n, act.Name, ret)
+		}
+		// And, since we can't really return an error to the process, but only to strace
+		// and, since this is more about killing it than anything else ...
+		// bye bye
+		c.cancel()
+	// Actually, log is the same as error,0 but ... it makes the intent clearer.
+	case "log":
+		if c.Log != nil {
+			fmt.Fprintf(c.Log, "%v act %v error %v\n", n, act.Name, ret)
+		}
+	}
+	return ret
+}
+
+// Run implements cmd.Run, filtering strace events
+// as created by AddActions. The slice can be empty, in which case the command
+// runs as normal.
+func (c *Cmd) Run() error {
+	if err := strace.Trace(c.Cmd, func(t strace.Task, r *strace.TraceRecord) error {
+		ret := c.handleEvent(t, r, c.events)
+		return ret
+	}); err != nil {
+		return fmt.Errorf("%v", err)
+	}
+	return nil
+}
+
+// AddActions creates an []event as defined by a possibly empty set of actions, and
+// installs it in the Cmd.
+func (c *Cmd) AddActions(actions ...string) error {
```

Comment:
can you document these filters in the function documentation?

Reply:
done.